air/src/proof.rs

```diff
@@ -1,102 +1,109 @@
+use winter_air::{proof::StarkProof, FieldExtension, ProofOptions as WinterProofOptions};
 
-// ExecutionProof
+// EXECUTION PROOF
 // ================================================================================================
 
-/// An execution proof with its metadata to define the security parameters and hasher primitive.
+/// A proof of correct execution of Miden VM.
+///
+/// The proof encodes the proof itself as well as STARK protocol parameters used to generate the
+/// proof. However, the proof does not contain public inputs needed to verify the proof.
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct ExecutionProof {
-    hasher: HashFunction,
     proof: StarkProof,
+    hash_fn: HashFunction,
 }
 
 impl ExecutionProof {
     // CONSTRUCTOR
     // --------------------------------------------------------------------------------------------
 
-    /// Creates a new instance of an execution proof from the used hash primitive and the STARK
-    /// proof.
-    pub const fn new(hasher: HashFunction, proof: StarkProof) -> Self {
-        Self { hasher, proof }
+    /// Creates a new instance of [ExecutionProof] from the specified STARK proof and hash
+    /// function.
+    pub const fn new(proof: StarkProof, hash_fn: HashFunction) -> Self {
+        Self { proof, hash_fn }
     }
 
     // PUBLIC ACCESSORS
     // --------------------------------------------------------------------------------------------
 
-    /// Returns the used hash primite to compute the proof.
-    pub const fn hasher(&self) -> HashFunction {
-        self.hasher
-    }
-
     /// Returns the underlying STARK proof.
-    pub const fn proof(&self) -> &StarkProof {
+    pub const fn stark_proof(&self) -> &StarkProof {
         &self.proof
     }
 
+    /// Returns the hash function used during proof generation process.
+    pub const fn hash_fn(&self) -> HashFunction {
+        self.hash_fn
+    }
+
+    /// Returns conjectured security level of this proof in bits.
+    pub fn security_level(&self) -> u32 {
+        match self.hash_fn {
+            HashFunction::Blake3_192 => self.proof.security_level::<Blake3_192>(true),
+            HashFunction::Blake3_256 => self.proof.security_level::<Blake3_192>(true),
+            HashFunction::Rpo256 => self.proof.security_level::<Blake3_192>(true),
+        }
+    }
+
     // SERIALIZATION / DESERIALIZATION
     // --------------------------------------------------------------------------------------------
 
     /// Serializes this proof into a vector of bytes.
     pub fn to_bytes(&self) -> Vec<u8> {
         let mut bytes = self.proof.to_bytes();
-        if bytes.is_empty() {
-            // TODO maybe error?
-            return vec![];
-        }
-        bytes.insert(0, self.hasher as u8);
+        assert!(!bytes.is_empty(), "invalid STARK proof");
+        // TODO: ideally we should write hash function into the proof first to avoid reallocations
```

Comment:
True, but I think we can achieve that only if we import the bytes writer from winter-utils